File: src/opensubtitles.ts
import { info, warn, error } from "./logger";

export async function searchSubtitles(imdbId: string, queries: string[] | undefined, env: any) {
  const queryCandidates = Array.isArray(queries) ? queries : queries ? [queries] : [];
  info("Searching OpenSubtitles for IMDb ID", { imdbId, queryCandidates });

  const apiKey = env.OPENSUBTITLES_API_KEY;
  if (!apiKey) {
    error("Missing OpenSubtitles API key");
    throw new Error("OpenSubtitles API key is not configured");
  }

  const actuallyQuery = async (query?: string) => {
    const params = new URLSearchParams({
      imdb_id: imdbId,
      languages: "en"
    });
    if (query) {
      params.set("query", query);
    }

    const res = await fetch(
      `https://api.opensubtitles.com/api/v1/subtitles?${params.toString()}`,
      {
        headers: {
          "Api-Key": apiKey,
          "User-Agent": "Stremio Hebrew Subtitles/1.0",
          "Accept": "application/json"
        }
      }
    );

    info("OpenSubtitles API response", {
      imdbId,
      query,
      status: res.status
    });

    const body = await res.text();
    let data: any;

    try {
      data = JSON.parse(body);
    } catch (parseError) {
      error("Failed to parse OpenSubtitles response", {
        imdbId,
        query,
        status: res.status,
        parseError,
        body
      });
      throw new Error("Invalid OpenSubtitles response");
    }

    if (!res.ok) {
      error("OpenSubtitles API error", {
        imdbId,
        query,
        status: res.status,
        response: data
      });
      throw new Error("OpenSubtitles API request failed");
    }

    const items: any[] = data.data || [];
    const itemSummaries = items.slice(0, 10).map((item) => ({
      id: item.id,
      release: item.attributes?.release,
      language: item.attributes?.language,
      download_count: item.attributes?.download_count,
      url: item.attributes?.url,
      file_id: item.attributes?.files?.[0]?.file_id ?? item.attributes?.file_id
    }));

    info("OpenSubtitles subtitle candidates", {
      imdbId,
      query,
      total: items.length,
      sample: itemSummaries
    });

    return items;
  };

  const triedQueries: any[] = [];
  if (queryCandidates.length === 0) {
    const items = await actuallyQuery();
    triedQueries.push({ query: undefined, count: items.length });
    info("OpenSubtitles search result", {
      imdbId,
      triedQueries
    });
    return items;
  }

  for (const query of queryCandidates) {
    const items = await actuallyQuery(query);
    triedQueries.push({ query, count: items.length });
    if (items.length > 0) {
      info("OpenSubtitles search result", {
        imdbId,
        triedQueries,
        selectedQuery: query
      });
      return items;
    }
  }

  info("OpenSubtitles search returned no results for any query", {
    imdbId,
    triedQueries
  });
  return [];
}

export async function resolveSubtitleDownloadUrl(item: any, env: any) {
  const apiKey = env.OPENSUBTITLES_API_KEY;
  if (!apiKey) {
    error("Missing OpenSubtitles API key for download resolution");
    throw new Error("OpenSubtitles API key is not configured");
  }

  const fileId =
    item.attributes?.files?.[0]?.file_id ??
    item.attributes?.file_id;

  if (!fileId) {
    const fallbackUrl = item.attributes?.url;
    info("No file_id available; falling back to attributes.url", {
      id: item.id,
      fallbackUrl
    });

    if (typeof fallbackUrl === "string" && fallbackUrl.length) {
      return fallbackUrl;
    }

    error("Subtitle item has no file_id or fallback URL", { item });
    throw new Error("Subtitle item has no downloadable source");
  }

  info("Resolving OpenSubtitles download link", {
    id: item.id,
    fileId
  });

  const downloadRes = await fetch("https://api.opensubtitles.com/api/v1/download", {
    method: "POST",
    headers: {
      "Api-Key": apiKey,
      "User-Agent": "Stremio Hebrew Subtitles/1.0",
      "Accept": "application/json",
      "Content-Type": "application/json"
    },
    body: JSON.stringify({ file_id: fileId })
  });

  const downloadBody = await downloadRes.text();
  let downloadData: any;

  try {
    downloadData = JSON.parse(downloadBody);
  } catch (parseError) {
    error("Failed to parse OpenSubtitles download response", {
      id: item.id,
      fileId,
      status: downloadRes.status,
      parseError,
      downloadBody
    });
    throw new Error("Invalid OpenSubtitles download response");
  }

  info("OpenSubtitles download response", {
    id: item.id,
    fileId,
    status: downloadRes.status,
    downloadData
  });

  if (!downloadRes.ok) {
    error("OpenSubtitles download endpoint failed", {
      id: item.id,
      fileId,
      status: downloadRes.status,
      response: downloadData
    });
    const fallbackUrl = item.attributes?.url;
    if (typeof fallbackUrl === "string" && fallbackUrl.length) {
      return fallbackUrl;
    }
    throw new Error("Failed to resolve OpenSubtitles download link");
  }

  const link =
    typeof downloadData?.link === "string"
      ? downloadData.link
      : typeof downloadData?.url === "string"
      ? downloadData.url
      : undefined;

  if (!link) {
    error("OpenSubtitles download response has no link", {
      id: item.id,
      fileId,
      downloadData
    });
    const fallbackUrl = item.attributes?.url;
    if (typeof fallbackUrl === "string" && fallbackUrl.length) {
      return fallbackUrl;
    }
    throw new Error("No download link returned from OpenSubtitles");
  }

  return link;
}

File: backend/opensubtitles.py
import logging
import re
from typing import Any, Dict, List, Optional

import httpx

logger = logging.getLogger("argos_subtitles.opensubtitles")

FILENAME_REMOVE_EXT = re.compile(r"\.json$|\.[^.]+$", re.IGNORECASE)


def filename_to_query(filename: str) -> str:
    return FILENAME_REMOVE_EXT.sub("", filename)


def build_filename_queries(filename: str) -> List[str]:
    base = filename_to_query(filename)
    normalized_separators = re.sub(r"[._,-]+", " ", base).strip()
    normalized_hyphens = re.sub(r"[-]+", " ", base).strip()

    candidates = [base]
    if normalized_separators and normalized_separators != base:
        candidates.append(normalized_separators)
    if normalized_hyphens and normalized_hyphens != base and normalized_hyphens != normalized_separators:
        candidates.append(normalized_hyphens)

    return list(dict.fromkeys(candidates))


def search_subtitles(imdb_id: str, filename: Optional[str], api_key: str) -> List[Dict[str, Any]]:
    if not api_key:
        raise RuntimeError("OpenSubtitles API key is not configured")

    queries = build_filename_queries(filename) if filename else [""]
    logger.info("Searching OpenSubtitles", {"imdb_id": imdb_id, "queries": queries})

    headers = {
        "Api-Key": api_key,
        "User-Agent": "Stremio Hebrew Subtitles/1.0",
        "Accept": "application/json"
    }

    def parse_response(res: httpx.Response) -> List[Dict[str, Any]]:
        try:
            data = res.json()
        except Exception as exc:
            logger.error("Failed to parse OpenSubtitles response", exc_info=exc)
            raise RuntimeError("Invalid OpenSubtitles response")

        if not res.is_success:
            logger.error("OpenSubtitles API error", {"status": res.status_code, "body": data})
            raise RuntimeError("OpenSubtitles API request failed")

        return data.get("data") or []

    for query in queries:
        params = {"imdb_id": imdb_id, "languages": "en"}
        if query:
            params["query"] = query

        response = httpx.get("https://api.opensubtitles.com/api/v1/subtitles", params=params, headers=headers, timeout=30.0)
        logger.info("OpenSubtitles search response", {"imdb_id": imdb_id, "query": query, "status": response.status_code})
        items = parse_response(response)
        if items:
            logger.info("OpenSubtitles found candidates", {"imdb_id": imdb_id, "query": query, "count": len(items)})
            return items

    logger.info("OpenSubtitles returned no results", {"imdb_id": imdb_id, "queries": queries})
    return []


def rank_subtitles(items: List[Dict[str, Any]]) -> Dict[str, Any]:
    if not items:
        raise RuntimeError("No subtitles found")

    ranked = sorted(items, key=lambda item: item.get("attributes", {}).get("download_count", 0), reverse=True)
    best = ranked[0]
    logger.info("Selected best subtitle candidate", {
        "id": best.get("id"),
        "release": best.get("attributes", {}).get("release"),
        "download_count": best.get("attributes", {}).get("download_count", 0)
    })
    return best


def resolve_subtitle_download_url(item: Dict[str, Any], api_key: str) -> str:
    if not api_key:
        raise RuntimeError("OpenSubtitles API key is not configured")

    attributes = item.get("attributes", {})
    file_id = None
    if isinstance(attributes.get("files"), list) and len(attributes["files"]) > 0:
        file_id = attributes["files"][0].get("file_id")

    if file_id is None:
        file_id = item.get("file_id")

    if file_id is None:
        fallback_url = attributes.get("url")
        if isinstance(fallback_url, str) and fallback_url:
            logger.info("Falling back to OpenSubtitles item URL", {"url": fallback_url})
            return fallback_url
        raise RuntimeError("Subtitle item has no downloadable source")

    logger.info("Resolving OpenSubtitles download link", {"file_id": file_id})
    response = httpx.post(
        "https://api.opensubtitles.com/api/v1/download",
        headers={
            "Api-Key": api_key,
            "User-Agent": "Stremio Hebrew Subtitles/1.0",
            "Accept": "application/json",
            "Content-Type": "application/json"
        },
        json={"file_id": file_id},
        timeout=30.0
    )

    try:
        data = response.json()
    except Exception as exc:
        logger.error("Failed to parse OpenSubtitles download response", exc_info=exc)
        raise RuntimeError("Invalid OpenSubtitles download response")

    if not response.is_success:
        logger.error("OpenSubtitles download endpoint failed", {"status": response.status_code, "body": data})
        fallback_url = attributes.get("url")
        if isinstance(fallback_url, str) and fallback_url:
            return fallback_url
        raise RuntimeError("Failed to resolve OpenSubtitles download link")

    link = data.get("link") or data.get("url")
    if not isinstance(link, str) or not link:
        logger.error("OpenSubtitles download response missing link", {"body": data})
        fallback_url = attributes.get("url")
        if isinstance(fallback_url, str) and fallback_url:
            return fallback_url
        raise RuntimeError("No download link returned from OpenSubtitles")

    return link
